CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account

marta- opened the pull request:
**Specs on Jira:**
* https://phenotips.atlassian.net/browse/CARDS-2369

**Testing:**
* start with --test
* create a new "Regular Expression Test" form and test the 4 flavours of text questions with regular expressions available there: 
  * enter values (valid and invalid)
  * save & view
  * re-edit
* edit the questionnaire and create another text question with a validation regexp, then test it

**Regression testing:**
* test text questions without regular expressions
* test other multiple choice questions:  "Multiple Choice Test", "Resource / Vocabulary Test", "Concept Recognition Test"

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account

sdumitriu reviewed (approved):
Works well enough. Some observations:
- Although not matching the regexp replaces the "press enter to add a new option" text with an error message, you can still press enter to "accept" the non-conformant test
- And in general, the warning doesn't prevent the user from moving on and saving the value; I guess that's not something to work on now, but we should work towards adding the INVALID status flag on the answer and prevent moving forward at least in the patient UI
- Saving an invalid text and re-editing the form does not warn about the non-conformant value, only after changing it does it show the error text
- It seems a bit aggressive to me that the error is displayed as soon as I start typing, I know it's not valid yet, give me a moment, I'm working on it

marta- commented:
Ok, let me make some tweaks

sashaandjic commented:
Not sure why the second field of the `Date Form Test` is behaving different then the first and third one.
![image](https://github.com/data-team-uhn/cards/assets/23243244/f2b316e0-8ddb-441f-998b-20a356919751)

https://github.com/data-team-uhn/cards/assets/23243244/51ddf7c6-2d00-468e-9653-84378897263f

marta- commented:
> Not sure why the second field of the `Date Form Test` is behaving different then the first and third one. I like the fact that it dies not complain right away, however once I input a valid format it clears the field.

Valid concerns, however they're unrelated to this PR. Please create a separate task. I'll add some testing instructions once I'm done (or give up on) addressing Sergiu's comments.

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account

* Better UX for multivalued questions with regexp validation
* Added a `liveValidation` flag to the question definition (validate as you type or when leaving the input)
* Highlight saved values that are invalid in "error" color
* Added backend regexp validation which sets the INVALID status flag
* Added a test questionnaire

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/internal/RegExpValueValidator.java:
```diff
@@ -0,0 +1,83 @@
+                // if any value does not match the pattern, set FLAG_INVALID to true
+                for (int i = 0; i < answerProp.count(); i++) {
+                    final String value = answerProp.getValue(Type.STRING, i);
+                    Pattern pattern = Pattern.compile(regexp);
```
Patterns should be compiled only once for better performance.

sdumitriu commented on modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/internal/RegExpValueValidator.java:
```diff
@@ -0,0 +1,83 @@
+                // This only works on text, nothing to do otherwise
+                return;
+            }
+            if (answer.hasProperty(PROP_VALUE)) {
```
Why not `if (question.hasProperty("validationRegexp") && answer.hasProperty(PROP_VALUE))`?

sdumitriu committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account
Minor codestyle improvements

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -231,9 +242,28 @@ function MultipleChoice(props) {
     setSelection((old) => unselect(old, id));
   }
 
-  let acceptEnteredOption = (overrideValue, overrideName) => {
```
I think overrideName was the one responsible for showing Yes|No instead of 1|0 for booleans.

sdumitriu commented:
Rebased on dev.

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account
Fixed regression: boolean-specific props are not getting forwarded to MultipleChoice

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account
Fixed javascript console warning introduced in d60f925 with the validattion error message as decriptions of entered options:
"react.development.js:209 Warning: Failed prop type: Invalid prop `children` of type `boolean` supplied to `FormattedText`, expected `string`."

sashaandjic commented:
Small complaint (probably out of scope); for single value fields, when in `view` mode replace `Please enter a valid ...` with `This is not a valid ...`
![image](https://github.com/data-team-uhn/cards/assets/23243244/7d74dbb6-a365-462f-8aa5-c5ef5dd57891)

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -53,7 +53,8 @@ const GHOST_SENTINEL = "custom-input";
   let { classes, customInput, customInputProps, existingAnswer, input, textbox, onUpdate, onChange, additionalInputProps, muiInputProps, naValue, noneOfTheAboveValue, error, questionName, ...rest } = props;
-  let { maxAnswers, minAnswers, displayMode, enableSeparatorDetection, validate, validationErrorText, liveValidation } = {...props, ...props.questionDefinition};
+  let { maxAnswers, minAnswers, displayMode, enableSeparatorDetection } = {...props.questionDefinition, ...props};
+  let { validate, validationErrorText, liveValidation } = {...props, ...props.questionDefinition};
```
The correct order is always `{...props.questionDefinition, ...props}`. We read the values from the question definition stored in the JCR, but we allow them to be overriden by the specific props passed from code. There are only 2 places where we do it in the other order, in `TextQuestion` and here, which I think is a mistake.

sdumitriu replied:
Oh, both instances are newly introduced in this PR. Why do you need to do it in this order and not the regular one?

marta- replied:
The  default `validationErrorText` is in `TextQuestion`'s `defaultProps`, and we don't want it (edit: it = the default) to overwrite whatever comes from the `questionDefinition`.

sdumitriu replied:
Why not? Or, why even bother reading it from `questionDefinition` if we don't want it?

sdumitriu replied:
Oh, I see, you mean that there is a default prop value. The standard way to do that is to use a default setting outside of the prop defaults, e.g. 

>   let { validate, validationErrorText, liveValidation } = {'validationErrorText': 'blabla', ...props.questionDefinition, ...props};

or

> = {...defaultValues, ...props.questionDefinition, ...props}

marta- replied:
> Why not? Or, why even bother reading it from questionDefinition if we don't want it?

I believe I wasn't clear. We want the `validationErrorText` from the `questionDefinition`, if it is specified. If none is specified, we want to use the default one. That is why on line 57 **first** we list `...props`, which contains the default, and **then** we list `...props.questionDefinition` which contains the admin-specified error text for that question which would overwrite the default.

If this workflow seems awkward, suggestions are welcome for better ways to specify the default error text.

marta- replied:
Refactored in e4fe26f.

marta- commented:
> Small complaint; for single value fields, replace `Please enter a valid ...` with `This is not a valid ...`.

I can change this in the settings of the test questionnaire. Keep in mind this text is a question-level setting, so it's up to whomever designs the questionnaire to use the appropriate wording for each case. FTR the default  when the error text isn't provided is "Invalid input".

sashaandjic commented:
tested
used `^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4}$` for phone number
![image](https://github.com/data-team-uhn/cards/assets/23243244/3ece1505-ea1b-4800-9f70-df1793034871)

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account

Updated the test questionnaire:
* Better validation error text for the single valued input
* Removed the validation error text from one of the question definitions to showcase the default error text

marta- committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account

Addressing comments: refactored prop passing from TextQuestion to MultipleChoice

marta- commented:
> > Small complaint; for single value fields, replace `Please enter a valid ...` with `This is not a valid ...`.
> 
> I can change this in the settings of the test questionnaire. Keep in mind this text is a question-level setting, so it's up to whomever designs the questionnaire to use the appropriate wording for each case. FTR the default when the error text isn't provided is "Invalid input".

Done in ac9e48f . Also removed the validation message from the second question to show how it defaults to "Invalid input".

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/Question.json:
```diff
@@ -17,6 +17,8 @@
             "minAnswers": "long",
             "maxAnswers": "long",
             "validationRegexp": "string",
+            "validationErrorText": "string",
+            "liveValidation": "boolean",
```
Todo: add these 2 new props to form.cnd as well @sdumitriu

sdumitriu committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account
Add the new properties to the CND

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/TextQuestion.jsx:
```diff
@@ -55,33 +55,47 @@ import AnswerComponentManager from "./AnswerComponentManager";
 //    displayMode={"input"}
-//    regexp={"[a-z]+"}
-//    errorText={"Please enter a lowercase input"}
+//    validationRegexp={"[a-z]+"}
```
```suggestion
//    validationRegexp={"^[a-z]+$"}
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/TextQuestion.jsx:
```diff
@@ -41,8 +41,8 @@ import AnswerComponentManager from "./AnswerComponentManager";
-//  regexp: String of a regular expression tested against the input
-//  errorText: String to display when the regexp is not matched
+//  validationRegexp: String of a regular expression tested against the input
+//  validationErrorText: String to display when the regexp is not matched
```
Except that right now, these aren't read from the props, just from the question definition. For consistency, they should also be read from props.

sashaandjic reviewed (approved):
tested

sdumitriu committed:
CARDS-2369: The validation regexp specified in the questionnaire designer for text questions is not taken into account
